### ISSUE TITLE
Weights for examples

### ISSUE THREAD

**Issue by eakbas** [vlfeat/matconvnet#49]
For extremely inbalanced datasets (i.e. needle in a haystack problem), how can I incorporate weights for examples? E.g., in the objective, I want to weigh positive examples, which are very rare, higher than the negative examples. What's the best way of doing this?

**Comment by lenck** [comment]
Hi, I think that this may be little bit complicated because the derivatives are collected over a batch inside the e.g. vl_nnconv.cu... So you would probably need to change the mex files and have adjustable `alpha` parameter in the sgemm operation for the backrpop (e.g. on line 820 in (vl_nnconv.cu)[https://github.com/vlfeat/matconvnet/blob/master/matlab/src/vl_nnconv.cu] but on many more places too...).

But if you want to just have a proof of concept, and you don't mind that the computation would take longer, you can pass batches of size 1 and then integrate the gradients in Matlab... Simplest would be to have your own version of `vl_simplenn` which would iterate over all images in a batch summing up the derivatives...

**Comment by lenck** [comment]
Hmm, just realised that I might have been wrong about this - you just need to change the softmaxloss function implementation to weight the loss per learning example... [feeling stupid]

**Comment by dbbert** [comment]
How would I go about changing the softmaxloss function to weigh the examples?

I would adapt lines 59-60 as follows, where the weights vector is based on the label vector c:

t = Xmax + log(sum(ex,3)) - reshape(X(c_), [sz(1:2) 1 sz(4)]) ;
t = t .* weights;
Y = sum(t(:)) / n ;

But what about the backward step?

**Comment by shenyu2013** [comment]
I also have this problem. After changing the softmaxloss function to weight the examples, how to change the backpropagation steps? 
I have tried it if nothing is done, that would cause NaNs and no regression.
